dsl_parser/functions.py

```diff
@@ -892,30 +860,98 @@ def handler(v, scope, context, path):
             if not isinstance(func, Function):
                 break
             previous_evaluated_value = evaluated_value
-            evaluated_value = getattr(func, evaluator)(**evaluator_kwargs)
+            evaluated_value = self.evaluate_function(func)
             if scanned and previous_evaluated_value == evaluated_value:
                 break
             scanned = True
         return evaluated_value
-    return handler
+
+    def evaluate_function(self, func):
+        return func.evaluate(self)
 
 
-def plan_evaluation_handler(plan):
-    return _handler('evaluate', plan=plan)
+class _PlanEvaluationHandler(_EvaluationHandler):
+    def __init__(self, plan, eager):
+        self._plan = plan
+        self._eager = eager
 
+    def evaluate_function(self, func):
+        if self._eager and func.func_eval_type in (HYBRID_FUNC, STATIC_FUNC):
+            return func.evaluate(self)
+        else:
+            if 'operation' in func.context:
+                func.context['operation']['has_intrinsic_functions'] = True
+            return func.raw
+
+    def get_input(self, input_name):
+        try:
+            return self._plan.inputs[input_name]
+        except KeyError:
+            raise exceptions.UnknownInputError(
+                "get_input function references an "
+                "unknown input '{0}'.".format(input_name))
+
+    def get_node(self, node_name):
+        for n in self._plan.node_templates:
+            if n['name'] == node_name:
+                return n
+        else:
+            raise KeyError('Node {0} does not exist'.format(node_name))
+
+
+class _RuntimeEvaluationHandler(_EvaluationHandler):
+    def __init__(self, storage):
+        self._storage = storage
+        self._nodes_cache = {}
+        self._node_to_node_instances = {}
+        self._node_instances_cache = {}
+        self._secrets_cache = {}
```

Comment:
I am not about caching entities that could change during the lifespan of a deployment like capabilities, node instance (in scaling policy, dep update will de-validate this?, secrets too)

Reply:
this cache lives for the duration of a single evaluate call, not for the duration of the whole deployment

and it worked before - this cache isn't new..